src/v/config/constraints.cc

```diff
@@ -0,0 +1,727 @@
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property) {
+    if (enabled == constraint_enabled_t::no) {
+        // A constraint that is turned off (i.e., no) means that the broker
+        // should not compare the topic property to the cluster one. This
+        // implies the the topic property automatically satisfies the
+        // constraint.
+        return true;
+    }
+
+    // A constraint with "enabled" flag means that the topic property must match
+    // the cluster one. A disabled/undefined topic value or cluster value could
+    // not match the other, this implies that the constraint is not satisfied.
+    if (!topic_val.has_optional_value() || !cluster_val) {
+        return false;
+    }
+
+    if (topic_val.value() != *cluster_val) {
+        vlog(
+          constraints_log.error,
+          "Constraints failure[does not match the cluster property {}]: "
+          "topic property {}.{}, value {}",
+          cluster_property,
+          topic(),
+          topic_property,
+          topic_val);
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Returns true if the topic-level value satisfies the constraint.
+ * \param topic_val: the topic value
+ * \param constraint: the constraint to evaluate
+ * \param property_name: name of the cluster-level property
+ * \param cluster_opt: the value from the cluster-level property
+ */
+template<typename T>
+bool validate_value(
+  const tristate<T>& topic_val,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    if (constraint.name == cluster_property) {
+        return ss::visit(
+          constraint.flags,
+          [&topic_val,
+           &topic,
+           &topic_property,
+           &cluster_property,
+           &cluster_opt](const constraint_enabled_t enabled) {
+              return matches_cluster_property_value(
+                topic_val,
+                cluster_opt,
+                enabled,
+                topic,
+                topic_property,
+                cluster_property);
+          },
+          [&topic_val, &topic, &topic_property](const auto range) {
+              return within_range(topic_val, range, topic, topic_property);
+          });
+    }
+    return false;
+}
+
+template<typename T>
+bool validate_value(
+  const std::optional<T>& topic_opt,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    auto tri = tristate<T>{topic_opt};
+
+    return validate_value(
+      tri, constraint, topic, topic_property, cluster_property, cluster_opt);
+}
+} // namespace
+
+bool topic_config_satisfies_constraint(
+  const cluster::topic_configuration& topic_cfg,
+  const constraint_t& constraint) {
```

Comment:
suggestion: Why not have the caller supply the actual topic config value and cluster config?  Something like:

```
template<typename T>
bool topic_config_satisifes_constraint(
  T & topic_config_val
  const property<T> &cluster_config)
```

This may need a few specializers maybe for `std::optional` topic config values

You can obtain the name of the cluster property via `cluster_config.name()` and with that you can look up the corresponding constraint, if it exists.  You may still need a map of sorts to get the corresponding topic property name

It looks like the caller already needs to know the name of the constraint in order to access it from the constraint map, right?

Reply:
> It looks like the caller already needs to know the name of the constraint in order to access it from the constraint map, right?

We do not need to know the name of the constraint in the general case. [Here](https://github.com/NyaliaLui/redpanda/commit/3065a563641b06ee3a91c2fa2b5b06dd466103b1#diff-2526100bdb158053e8146a22623b4fa084c33b8730e47715c43028636902235aR303-R317) is an example of what using this API looks like for create_topics. Note that `apply_constraint` was renamed to `valid_topic_config` (I'm still working on the naming but that is a non-blocker for this PR). Notice that the broker iterates over every constraint and then checks the topic configuration against the constraint.

For those APIs that need a specific constraint, a search by name is simple. [Here](https://github.com/NyaliaLui/redpanda/commit/3065a563641b06ee3a91c2fa2b5b06dd466103b1#diff-215eea521e18fe56d728c5bf64ada99aa4790c0d7f0816b660a2b6f156b99fd3R286-R301) is an example with CreatePartitions.

> Why not have the caller supply the actual topic config value and cluster config?

Having the caller supply the actual topic config value would re-create a similar structure to what is already included. [here](https://github.com/NyaliaLui/redpanda/commit/19039107db4b229a1135049c327bd77f6d15566d#diff-c618e718c29ebdabd2aadde150ff01dc2b6750369c9f798dec0c1de182791459R364-R378) is an example at the callsite.

> You can obtain the name of the cluster property via cluster_config.name() and with that you can look up the corresponding constraint, if it exists.

See the above comment about constraint search.